### PR TITLE
ARROW-7531: [C++] Reduce header cost

### PR DESCRIPTION
**Pull request by pitrou**
Reduce compilation times by ~9% in debug mode.

In particular, make sure the concrete Array and Builder classes are only included when required.
(`arrow/array/builder_dict.h` is pathologic, because it includes the entire memo table and xxhash implementation).

### CODE REVIEW

**Comment by pitrou** [comment]
Running [Clang Build Analyzer](https://github.com/aras-p/ClangBuildAnalyzer) on git master:
```
**** Time summary:
Compilation (368 times):
  Parsing (frontend):          440.0 s
  Codegen & opts (backend):    123.6 s

```
and on this PR:
```
**** Time summary:
Compilation (369 times):
  Parsing (frontend):          383.3 s
  Codegen & opts (backend):    124.1 s
```

Note that the times measured are bit [unreliable](https://github.com/aras-p/ClangBuildAnalyzer/issues/28).

**Comment by pitrou** [comment]
This is the typical output of analyzing total header cost:
https://gist.github.com/pitrou/35e53b57bf0f0f57ec40a4d3721c376d

**Comment by pitrou** [comment]
@wesm Should `api.h` include all the concrete Array and Builder classes?

**Comment by wesm** [comment]
@pitrou I think it probably should. We haven't been maintaining the api.h files very diligently though

File: cpp/src/arrow/array_all.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "arrow/array.h"                  // IWYU pragma: export
```
Comment (wesm) [inline cpp/src/arrow/array_all.h]: It seems like `array_all.h` should be `array.h` and what is here array.h should be array/array_base.h or something

Comment (pitrou) [reply cpp/src/arrow/array_all.h]: If so, then it means there's not much benefit in splitting header files (except possibly readability). Everyone will intuitively include `array.h` and `builder.h`.

**Comment by wesm** [comment]
> If so, then it means there's not much benefit in splitting header files (except possibly readability). Everyone will intuitively include array.h and builder.h.

Well, there is the public API and then there is what headers we include in internal modules. It seems like the spirit of this PR is to reduce the amount of superfluously included code, which means we should split up array.h so each part can be included piecemeal. See http://issues.apache.org/jira/browse/ARROW-6052

**Comment by pitrou** [comment]
I see that the C Glib bindings also include `arrow/api.h`...

File: cpp/src/arrow/array_all.h (hunk 1 above)
Comment (emkornfield) [reply cpp/src/arrow/array_all.h]: Would maybe not including this these files be too onerous for developers?

Another advantage of more fine-grained header files is the potentially the elimination of "not real cycles" between compilation units I mentioned on the bazel thread.

Comment (pitrou) [reply cpp/src/arrow/array_all.h]: My experience in Python would be that finer-grained includes / imports increase the likelihood of cycles. For example, `array_list.h` and `array_binary.h` might have a cyclic dependency between each other while a monolithic `array.h` would not have a cycle onto itself (of course, this is a fictional example).

**Comment by wesm** [comment]
It seems to me that we should keep `arrow/array.h` and `arrow/builder.h` as batteries-included public APIs for downstream users of the project, but these should not use IWYU export (rather simply `keep`) so that IWYU will protest if an internal module includes `arrow/array.h` instead of more simply for example `arrow/array/array_base.h`. I think the objective should be to keep things orderly and minimal internally within the C++ codebase? 

IWYU is a pain to keep well-configured to not spew out red herrings but it does help maintain some cleanliness with the includes if it is inspected frequently. Since we build cross-platform I don't think having a 100% clean IWYU build is going to be practical (it can be if you are Linux-only)

**Comment by pitrou** [comment]
It would be nice if we had a Docker container for IWYU. Last time I tried to run it, my terminal was overwhelmed with warnings. Generally, I'm not sure I understand what it's doing.

**Comment by wesm** [comment]
We used to have a docker-compose target for IWYU. I'm not sure if it still works @kszucs

**Comment by kszucs** [comment]
I also had difficult time with IWYU but didn't tune its configuration. IWYU is still installed in the lint image, and archery theoretically supports to run it, so the following should work:

```bash
docker-compose run --rm ubuntu-lint /bin/bash -c "pip install -e /arrow/dev/archery && archery lint --with-iwyu true"
```

cc @fsaintjacques

**Comment by fsaintjacques** [comment]
@pitrou close or rebase?

**Comment by pitrou** [comment]
Given the diffusion of changes accross the codebase, rebasing this wholesale would probably be painful. A better strategy would probably to retry and do some of the changes one by one, in separate PRs. I should try to do that someday, but this is also low-priority.

**Comment by wesm** [comment]
There's a lot of good changes here. I agree that starting fresh and making smaller PRs is the path forward. I'm going to close this for now if that's OK

**Pull request closed without merging**